internal/proxy/publish_handler_test.go

```diff
@@ -1 +1,206 @@
 package proxy
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/centrifugal/centrifugo/v3/internal/rule"
+
+	"github.com/centrifugal/centrifuge"
+	"github.com/centrifugal/centrifugo/v3/internal/proxyproto"
+	"github.com/centrifugal/centrifugo/v3/internal/tools"
+)
+
+type publishHandlerTestDepsConfig struct {
+	publishProxyHandler *PublishHandler
+	transport           *tools.TestTransport
+}
+
+func newPublishHandlerTestDepsConfig(proxyEndpoint string) publishHandlerTestDepsConfig {
+	proxyCfg := Config{
+		HTTPConfig: HTTPConfig{
+			Encoder: &proxyproto.JSONEncoder{},
+			Decoder: &proxyproto.JSONDecoder{},
+		},
+		PublishEndpoint: proxyEndpoint,
+	}
+
+	connectProxy, err := NewHTTPPublishProxy(
+		proxyEndpoint,
+		proxyCfg,
+	)
+	if err != nil {
+		log.Fatalln("could not create http publish proxy: ", err)
+	}
+
+	publishProxyHandler := NewPublishHandler(PublishHandlerConfig{
+		Proxy: connectProxy,
+	})
+
+	return publishHandlerTestDepsConfig{
+		publishProxyHandler: publishProxyHandler,
+		transport:           tools.NewTestTransport(),
+	}
+}
+
+func TestHandlePublishWithResult(t *testing.T) {
+	node := tools.NodeWithMemoryEngineNoHandlers()
+	defer func() { _ = node.Shutdown(context.Background()) }()
+
+	custData := "test"
+	custDataB64 := base64.StdEncoding.EncodeToString([]byte(custData))
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/publish", func(w http.ResponseWriter, req *http.Request) {
+		_, _ = w.Write([]byte(fmt.Sprintf(`{"result": {"user": "56", "expire_at": 1565436268, "b64data": "%s"}}`, custDataB64)))
+	})
+	server := httptest.NewServer(mux)
+	defer server.Close()
+
+	testDepsCfg := newPublishHandlerTestDepsConfig(server.URL + "/publish")
+	publishHandler := testDepsCfg.publishProxyHandler.Handle(node)
+
+	client, closeFn, err := centrifuge.NewClient(context.Background(), node, testDepsCfg.transport)
+	require.NoError(t, err)
+	defer func() { _ = closeFn() }()
+
+	publishEvent := centrifuge.PublishEvent{}
+	chOpts := rule.ChannelOptions{
+		HistoryTTL:  1,
+		HistorySize: 1,
+	}
+
+	publishReply, err := publishHandler(client, publishEvent, chOpts)
+	require.NoError(t, err)
+	require.Equal(t, uint64(1), publishReply.Result.Offset)
```

Comment:
Could you plz also check non-empty Result.Epoch here?

Reply:
done